### PR TITLE
create def ids for impl traits during ast lowering

### PR DESCRIPTION
**Pull request by spastorino**
r? @cjgillot

### CODE REVIEW

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 1]
```diff
@@ -1453,7 +1461,23 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
         // frequently opened issues show.
         let opaque_ty_span = self.mark_span_with_reason(DesugaringKind::OpaqueTy, span, None);
 
-        let opaque_ty_def_id = self.local_def_id(opaque_ty_node_id);
+        let opaque_ty_def_id = match origin {
+            hir::OpaqueTyOrigin::TyAlias(def_id) => {
+                if let Some(def_id) = self.opt_local_def_id(opaque_ty_node_id) {
```
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/lib.rs:1466]: I what case would this ever be `Some`?

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 2]
```diff
@@ -1453,7 +1461,23 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
...
+                    def_id
+                } else {
+                    self.create_def(def_id, opaque_ty_node_id, DefPathData::ImplTrait)
+                }
+            }
+            hir::OpaqueTyOrigin::FnReturn(fn_def_id) => {
+                if let Some(def_id) = self.opt_local_def_id(opaque_ty_node_id) {
```
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/lib.rs:1473]: I what case would this ever be `Some`?

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 3]
```diff
@@ -1351,15 +1351,23 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
...
+                        .lower_opaque_impl_trait(
+                            span,
+                            hir::OpaqueTyOrigin::TyAlias(*parent_def_id),
+                            def_node_id,
+                            bounds,
+                            false,
+                            itctx,
+                        ),
                     ImplTraitContext::Universal => {
+                        if self.opt_local_def_id(def_node_id).is_none() {
```
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/lib.rs:1364]: I what case would this ever be `Some`?

File: src/test/ui/impl-trait/nested-return-type2-tait3.rs [hunk 4]
```diff
@@ -24,8 +24,7 @@ type Traitable = impl Trait<Assoc = impl Send>;
 // type does not implement `Duh`, even if its hidden type does. So we error out.
 fn foo() -> Traitable {
     || 42
-    //~^ ERROR `impl Send: Duh` is not satisfied
+    //~^ ERROR the trait bound `Traitable::{opaque#1}: Duh` is not satisfied
```
Comment (cjgillot) [inline src/test/ui/impl-trait/nested-return-type2-tait3.rs:27]: The path here is wrong. `impl Send` should be `Traitable::{opaque#0}::{opaque#0}`.

File: compiler/rustc_ast_lowering/src/lib.rs (hunk 1 above)
Comment (spastorino) [reply compiler/rustc_ast_lowering/src/lib.rs:1466]: The 3 cases are coming from ...

```
stack backtrace:
   0: rust_begin_unwind
             at ./library/std/src/panicking.rs:556:5
   1: core::panicking::panic_fmt
             at ./library/core/src/panicking.rs:142:14
   2: <rustc_ast_lowering::LoweringContext>::lower_opaque_impl_trait
   3: <rustc_ast_lowering::LoweringContext>::lower_ty_direct
   4: <rustc_ast_lowering::LoweringContext>::lower_ty
             at ./compiler/rustc_ast_lowering/src/lib.rs:1197:26
   5: <rustc_ast_lowering::LoweringContext>::lower_assoc_ty_constraint::{closure#0}
             at ./compiler/rustc_ast_lowering/src/lib.rs:1079:34
   6: <rustc_ast_lowering::LoweringContext>::with_dyn_type_scope::<rustc_hir::hir::TypeBindingKind, <rustc_ast_lowering::LoweringContext>::lower_assoc_ty_constraint::{closure#0}>
             at ./compiler/rustc_ast_lowering/src/lib.rs:868:22
   7: <rustc_ast_lowering::LoweringContext>::lower_assoc_ty_constraint
             at ./compiler/rustc_ast_lowering/src/lib.rs:1071:21
   8: <rustc_ast_lowering::LoweringContext>::lower_angle_bracketed_parameter_data::{closure#2}
             at ./compiler/rustc_ast_lowering/src/path.rs:335:54
   9: <&mut <rustc_ast_lowering::LoweringContext>::lower_angle_bracketed_parameter_data::{closure#2} as core::ops::function::FnMut<(&rustc_ast::ast::AngleBracketedArg,)>>::call_mut
```

File: src/test/ui/impl-trait/nested-return-type2-tait3.rs (hunk 4 above)
Comment (spastorino) [reply src/test/ui/impl-trait/nested-return-type2-tait3.rs:27]: I'm not sure why this happen but I guess we need to call with_hir_id_owner and wrap some of the lowering code that's not currently wrapped?.

File: compiler/rustc_ast_lowering/src/lib.rs (hunk 2 above)
Comment (spastorino) [reply compiler/rustc_ast_lowering/src/lib.rs:1473]: Same as https://github.com/rust-lang/rust/pull/102483#discussion_r984017996

File: compiler/rustc_ast_lowering/src/lib.rs (hunk 3 above)
Comment (spastorino) [reply compiler/rustc_ast_lowering/src/lib.rs:1364]: Same as https://github.com/rust-lang/rust/pull/102483#discussion_r984017996

File: compiler/rustc_ast_lowering/src/lib.rs (hunk 1 above)
Comment (spastorino) [reply compiler/rustc_ast_lowering/src/lib.rs]: Ok, this is fixed now. Have seen that we are creating a def id for the impl trait in lower_assoc_ty_contraint

**Comment by spastorino** [comment]
It turned out that this was way simpler to what I was trying to do.

@cjgillot I think it's ready unless there's something outside ui tests that need to be blessed because def_ids ordering/numbers have changed now.

**Comment by cjgillot** [comment]
This looks like a bug in the stable hasher.  This branch is only used for item-relative spans (`-Zincremental-relative-spans`).  Using a `unwrap_or(DUMMY_SP)` is safe: span hashing won't try to make this a relative span.
The better solution would ofc to find a way to modify `resolutions(()).source_span`.

**Comment by spastorino** [comment]
@cjgillot now should be ready

**Pull request merged**